samcli/lib/samlib/resource_metadata_normalizer.py

```diff
@@ -44,18 +47,21 @@ def normalize(template_dict):
 
         for logical_id, resource in resources.items():
             resource_metadata = resource.get(METADATA_KEY, {})
-            asset_property = resource_metadata.get(ASSET_PROPERTY_METADATA_KEY)
-
-            if asset_property == IMAGE_ASSET_PROPERTY:
-                asset_metadata = ResourceMetadataNormalizer._extract_image_asset_metadata(resource_metadata)
-                ResourceMetadataNormalizer._update_resource_metadata(resource_metadata, asset_metadata)
-                # For image-type functions, the asset path is expected to be the name of the Docker image.
-                # When building, we set the name of the image to be the logical id of the function.
-                asset_path = logical_id.lower()
-            else:
-                asset_path = resource_metadata.get(ASSET_PATH_METADATA_KEY)
-
-            ResourceMetadataNormalizer._replace_property(asset_property, asset_path, resource, logical_id)
+            is_normalized = resource_metadata.get(SAM_IS_NORMALIZED, False)
+            if not is_normalized:
+                asset_property = resource_metadata.get(ASSET_PROPERTY_METADATA_KEY)
+                if asset_property == IMAGE_ASSET_PROPERTY:
+                    asset_metadata = ResourceMetadataNormalizer._extract_image_asset_metadata(resource_metadata)
+                    ResourceMetadataNormalizer._update_resource_metadata(resource_metadata, asset_metadata)
+                    # For image-type functions, the asset path is expected to be the name of the Docker image.
+                    # When building, we set the name of the image to be the logical id of the function.
+                    asset_path = logical_id.lower()
+                else:
+                    asset_path = resource_metadata.get(ASSET_PATH_METADATA_KEY)
```

Comment:
This for loop will go through all resources in the template and update their `asset_path`. Is this going to work only for function types or for all resource types?

Reply:
this loop is for all resources that has the CDK path metadata, so generally CDK will add the path metadata and set the resource property that will be replaced with the CDK path metadata value. This applied for any CDK L2 constructs that use assets